### PR TITLE
Genesis trigger

### PR DESCRIPTION
**Pull request by shayzluf**
[Part of] #2891 

---

# Genesis trigger

spec 0.8 change introduced new way to determine beacon chain genesis.

https://github.com/ethereum/eth2.0-specs/blob/dev/specs/core/0_beacon-chain.md#genesis

### CODE REVIEW

**Comment by shayzluf** [comment]
still work to be done on aligning is_valid_genesis_state to spec. and resolving proof mismatch in SetupInitialDeposits

File: beacon-chain/powchain/log_processing.go [hunk 1]
```diff
@@ -33,6 +54,14 @@ func (w *Web3Service) ProcessLog(depositLog gethTypes.Log) {
 	// Process logs according to their event signature.
 	if depositLog.Topics[0] == hashutil.HashKeccak256(depositEventSignature) {
 		w.ProcessDepositLog(depositLog)
+		if !w.chainStarted {
+			triggered := w.isValidGenesisState(uint64(len(w.chainStartDeposits)), 0)
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]: you are passing `0` as `currentTime`, how will the chain ever start?

File: beacon-chain/powchain/log_processing.go [hunk 2]
```diff
@@ -33,6 +54,14 @@ func (w *Web3Service) ProcessLog(depositLog gethTypes.Log) {
...
+			if triggered {
+				timeStamp := uint64(w.blockTime.Unix())
+				w.eth2GenesisTime = timeStamp - timeStamp%params.BeaconConfig().SecondsPerDay + 2*params.BeaconConfig().SecondsPerDay
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]: always a good idea to add a comment to explain complex math

File: beacon-chain/powchain/log_processing.go [hunk 3]
```diff
@@ -120,6 +149,44 @@ func (w *Web3Service) ProcessDepositLog(depositLog gethTypes.Log) {
 	}
 }
 
+// ProcessChainStart processes the log which had been received from
+// the ETH1.0 chain by trying to determine when to start the beacon chain.
+func (w *Web3Service) ProcessChainStart(genesisTime uint64) {
+	chainStartCount.Inc()
+
+	// w.chainStartETH1Data = &pb.Eth1Data{
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]: what is this?

File: beacon-chain/powchain/log_processing.go [hunk 4]
```diff
@@ -120,6 +149,44 @@ func (w *Web3Service) ProcessDepositLog(depositLog gethTypes.Log) {
...
+	// 	BlockHash:   depositLog.BlockHash[:],
+	// 	DepositRoot: chainStartDepositRoot[:],
+	// }
+
+	w.chainStarted = true
+	//w.depositRoot = chainStartDepositRoot[:]
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]: this too?

File: shared/params/config.go [hunk 5]
```diff
@@ -55,7 +56,7 @@ type BeaconChainConfig struct {
 	EpochsPerHistoricalVector uint64 `yaml:"EPOCHS_PER_HISTORICAL_VECTOR"` // EpochsPerHistoricalVector defines max length in epoch to store old historical stats in beacon state.
 	EpochsPerSlashingsVector  uint64 `yaml:"EPOCHS_PER_SLASHINGS_VECTOR"`  // EpochsPerSlashingsVector defines max length in epoch to store old stats to recompute slashing witness.
 	HistoricalRootsLimit      uint64 `yaml:"HISTORICAL_ROOTS_LIMIT"`       // HistoricalRootsLimit the define max historical roots can be saved in state before roll over.
-	ValidatorsLimit           uint64 `yaml:"VALIDATOR_REGISTRY_LIMIT"`     // ValidatorsLimit defines the upper bound of validators can participate in eth2.
+	ValidatorRagistryLimit    uint64 `yaml:"VALIDATOR_REGISTRY_LIMIT"`     // ValidatorRagistryLimit defines the upper bound of validators can participate in eth2.
```
Comment (terencechain) [inline shared/params/config.go]: spelled registry wrong

File: beacon-chain/powchain/log_processing.go [hunk 6]
```diff
@@ -13,18 +14,38 @@ import (
...
 )
 
 var (
 	depositEventSignature = []byte("DepositEvent(bytes,bytes,bytes,bytes,bytes)")
 )
 
+// isValidGenesisState gets called whenever there's a deposit event,
+// it checks whether there's enough effective balance to trigger and
+// if the minimum genesis time arrived already.
+// Spec pseudocode definition:
```
Comment (terencechain) [inline beacon-chain/powchain/log_processing.go]:
```suggestion

// Spec pseudocode definition:
```

File: beacon-chain/powchain/log_processing.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/powchain/log_processing.go]: missed it.
thanks

File: beacon-chain/powchain/log_processing.go (hunk 6 above)
Comment (shayzluf) [reply beacon-chain/powchain/log_processing.go]: thanks

File: shared/params/config.go (hunk 5 above)
Comment (shayzluf) [reply shared/params/config.go]: fixed. thanks

File: beacon-chain/powchain/log_processing.go (hunk 3 above)
Comment (shayzluf) [reply beacon-chain/powchain/log_processing.go]: removed. it is being initialized in genesis state

File: beacon-chain/powchain/log_processing.go (hunk 2 above)
Comment (shayzluf) [reply beacon-chain/powchain/log_processing.go]: added comment

**Pull request merged**